gossip/scrambler_new/tx_scrambler_test.go

```diff
@@ -0,0 +1,544 @@
+					hash:     common.Hash{3},
+					sender:   common.Address{3},
+					nonce:    3,
+					gasPrice: big.NewInt(3),
+				},
+				&dummyScramblerEntry{
+					hash:     common.Hash{4},
+					sender:   common.Address{2},
+					nonce:    2,
+					gasPrice: big.NewInt(4),
+				},
+				&dummyScramblerEntry{
+					hash:     common.Hash{5},
+					sender:   common.Address{1},
+					nonce:    1,
+					gasPrice: big.NewInt(5),
+				},
+			},
+		},
+		{
+			name: "repeated addresses, nonces and gas prices",
+			entries: []ScramblerEntry{
+				&dummyScramblerEntry{
+					hash:     common.Hash{1},
+					sender:   common.Address{1},
+					nonce:    1,
+					gasPrice: big.NewInt(1),
+				},
+				&dummyScramblerEntry{
+					hash:     common.Hash{2},
+					sender:   common.Address{2},
+					nonce:    2,
+					gasPrice: big.NewInt(2),
+				},
+				&dummyScramblerEntry{
+					hash:     common.Hash{3},
+					sender:   common.Address{3},
+					nonce:    3,
+					gasPrice: big.NewInt(3),
+				},
+				&dummyScramblerEntry{
+					hash:     common.Hash{4},
+					sender:   common.Address{2},
+					nonce:    2,
+					gasPrice: big.NewInt(2),
+				},
+				&dummyScramblerEntry{
+					hash:     common.Hash{5},
+					sender:   common.Address{1},
+					nonce:    1,
+					gasPrice: big.NewInt(1),
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			res1 := test.entries
+			res2 := slices.Clone(res1)
+			// shuffle one array
+			shuffleEntries(res2)
+
+			res1 = scrambleTransactions(42, res1)
+			res2 = scrambleTransactions(42, res2)
+			if slices.CompareFunc(res1, res2, compareFunc) != 0 {
+				t.Error("slices have different order - algorithm is not deterministic")
+			}
+		})
+	}
+}
+
+func TestTxScrambler_FilterAndOrderTransactions_RandomInput(t *testing.T) {
+	// this tests these input sizes:
+	// 1, 4, 16, 64, 256, 1024
+	for i := int64(1); i <= 1024; i = i * 4 {
+		entries := createRandomScramblerEntries(i)
+		copy := slices.Clone(entries)
+		shuffleEntries(copy)
+		entries = scrambleTransactions(42, entries)
+		copy = scrambleTransactions(42, copy)
+		if slices.CompareFunc(entries, copy, compareFunc) != 0 {
+			t.Error("slices have different order - algorithm is not deterministic")
+		}
+	}
+}
+
+// Permutations and pseudo random number generation
+
+func TestPseudoRandomNumberGenerator_OutputIsEquidistributed(t *testing.T) {
+	rnd := xorShiftStar{}
+	rnd.seed(42)
+	n := 1000000
+
+	seenXor := map[uint64]struct{}{}
+	seenRand := map[uint64]struct{}{}
+
+	for range n {
+		seenXor[rnd.randN(uint64(n))] = struct{}{}
+		seenRand[rnd.randN(uint64(n))] = struct{}{}
+	}
+
+	if len(seenXor) < len(seenRand)*99/100 {
+		t.Fatalf("Homemade rand is not within 1 percent of rand library: %d vs. %d", len(seenXor), len(seenRand))
+	}
+}
+func TestScrambler_shuffleSendersReturnsDifferentPermutations(t *testing.T) {
+	n := 8
+	numPermutations := fac(n)
+	seeds := make([]int, numPermutations)
+	for i := range seeds {
+		seeds[i] = i
+	}
```

Comment:
You create a list here to produce the numbers [0,1,...]; you can get this simpler if you use
```
for seed := range fac(n) {
```
below. Right?